File: src/learning_gem5/part2/hello_object.cc
#include "learning_gem5/part2/hello_object.hh"

#include "debug/HelloExample.hh"

namespace gem5
{

HelloObject::HelloObject(const HelloObjectParams &params)
    : SimObject(params),
      // event(*this), //不同版本
      event([this] { processEvent(); }, name()),
      myName(params.name), // 将参数对象中的name存储到成员变量myName中
      latency(params.time_to_wait),
      timesLeft(params.number_of_fires)
{
    DPRINTF(HelloExample, "Created the hello object with the name %s\n",
            myName.c_str());
} // 将声明的参数传递给C++类的构造函数

/*HelloObject::HelloObject(const HelloObjectParams &params)
    : SimObject(params),
      event([this] { processEvent(); }, name()),
      latency(100),
      timesLeft(10)
{
    DPRINTF(HelloExample, "Created the hello object\n");
} // 在构造函数中添加两个变量的默认值*/

/*HelloObject::HelloObject(const HelloObjectParams &params) :
    SimObject(params)
{
    // std::cout << "Hello World! From a SimObject!" << std::endl;
    DPRINTF(HelloExample, "Create the hello object\n");
    // 输出调试信息, 包括DPRINTF执行时的当前时钟周期,
    // 调用DPRINTF的SimObject名称 以及传递的字符串
    // 该函数接收两个参数: 1. 一个无参数无返回值的函数 2. 事件的名称,
    // 使用name函数获取 this放入lambda表达式中,
    // 以便在事件触发时调用processEvent函数
}*/

void
HelloObject::processEvent()
{
    timesLeft--;
    DPRINTF(HelloExample, "Hello world! Processing the event! %d left\n",
            timesLeft);

    if (timesLeft <= 0) {
        DPRINTF(HelloExample, "Done firing!\n");
    } else {
        schedule(event, curTick() + latency);
    }
}
// 定义进程函数的实现, 只需打印一些内容用作调试
// 更新函数

void
HelloObject::startup()
{
    // schedule(event, 100);
    schedule(event, latency); // 更新函数
}
// 安排事件的执行时间, startup()函数在模拟器启动时被调用,
// 这里安排事件在100周期(可以使用显式时间)后执行
} // namespace gem5

File: src/learning_gem5/part2/HelloObject.py
"""创建一个新的简单模拟对象, 先新建一个Python类"""

from m5.params import *
from m5.SimObject import SimObject


class HelloObject(SimObject):
    type = "HelloObject"  # 要封装的C++类
    cxx_header = "learning_gem5/part2/hello_object.hh"  # 包含C++类的头文件声明
    cxx_class = "gem5::HelloObject"  # 一个属性, 用于指定新创建的SimObject类声明在gem5命名空间中

    # 为HelloObject添加更多参数: 延迟和触发事件次数的参数
    # gem5 Python接口可以将参数从Python传递给C++对象
    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(
        1, "Number of times to fire the event before" " goodbye"
    )
    # Param类型声明一个参数, 常见的有Int Float String Bool等, 还可以使用自定义的Param类型, 例如Latency
    # 每个参数都需要一个默认值(不指定时为1)和一个描述字符串, 这些信息在gem5的文档中会被使用
    # Latency不是内置类型, 将参数转化为模拟器的滴答数, 默认是1皮秒, 也有其他参数: Percent, Cycles, MemorySize等
